### PR TITLE
[REF] (*_)website_sale(_*): streamline address management

### PR DESCRIPTION
**Pull request by Feyensv**
**[REF] (\*\_)website_sale(\_\*): streamline address management**

This commit aims to improve the address management of the eCommerce checkout for:
- address choice (`/shop/checkout` page)
- address creation and update (`/shop/address` page)

The approach is mostly centered around the controllers for now; future commits will handle other aspects of this logic.

The first aim of this refactor is to clarify and simplify as much as possible the code of those routes to ease future tasks, improve maintainability, etc.

Main changes are:
- Split the `/shop/address` page logic between 2 routes:
  - `/shop/address`: display the address form to the user
  - `/shop/address/submit`: submit the address form data
- Factorize shared code into util methods.
- Simplify and improve existing methods and their overrides:
  - Rely on recordsets rather than record ids.
  - Group the full pre-processing inside `_values_preprocess`.
- Order routes to follow the standard checkout flow.
- Apply code guidelines to improve readability.
- Use the simplified 'mode' everywhere, either billing or shipping, to know whether we are modifying an existing address. The presence of a partner argument distinguishes the cases where the address must be edited or created.
- Stop rendering the page on form submission (significant speed up).
- Delegate parts of the input validation to the browser to remove useless requests for invalid/incomplete addresses.
- Correctly pair labels and inputs (the labels for attributes were the input's `name` and not the input's `id`).

---

**[IMP] website_sale: simplify pricelist management**

Now that we have a dedicated cart method to update the cart address(es), it can be used in `sale_get_order` to simplify the method.

While on that, harmonize and simplify a bit the API related to pricelists updates on the eCommerce.

---

**[IMP] website_sale: harmonize address creation flows**

The flows are: the standard address creation flow, and the express checkout flow.

---

task-3682746

See also:
- https://github.com/odoo/enterprise/pull/54239

### CODE REVIEW

**Review by chevalierv — COMMENTED** [review]
Quick pass :duck:

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -1416,15 +1557,15 @@ def _create_or_edit_partner(self, partner_details, edit=False, **custom_values):
         :return int: The id of the partner created or edited
         """
         request.update_env(context=request.website.env.context)
-        values = self.values_preprocess(partner_details)
+        sanitized_values, _side_values = self._values_preprocess(partner_details)
 
+        # TODO VCR use _fill_partner_creation_values ?
+        # does it really make sense to update all those fields on an existing partner
+        # instead of only using lang/company/team/user/... in partner creation vals ?
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/controllers/main.py]: Done in the harmonize address creation commit

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -896,14 +907,17 @@ def _get_cart_notification_information(self, order, line_ids):
     # Checkout
     # ------------------------------------------------------
 
-    def checkout_check_address(self, order):
-        partner_invoice = order.partner_invoice_id
+    def checkout_check_address(self, order_sudo):
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/controllers/main.py]: Done :sunglasses:

**Review by vchu-odoo — COMMENTED** [review]
Hello,

That is a huge and great work!
Here are some small comments so far.

File: addons/website_sale/controllers/main.py [hunk 3]
```diff
@@ -983,340 +1037,501 @@ def checkout_values(self, order, **kw):
...
+                order_sudo.partner_shipping_id,
+            ):
+                # Only check partner existence if it's not already linked to the SO
+                partner_sudo = partner_sudo.exists()
+
+        if partner_sudo and not mode:
+            if partner_id == order_sudo.partner_invoice_id.id:
+                mode = 'billing'
+            elif partner_id == order_sudo.partner_shipping_id.id:
+                mode = 'shipping'
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/controllers/main.py:1533]: IIRC, the `billing` address requires more data than the `shipping` one, so we must make sure that the billing mode takes precedence

**Review by vchu-odoo — APPROVED** [review]
Great work!
LGTM :+1:

**Comment by AntoineVDV** [comment]
Sorry for the ping on the JS team :sweat_smile: This PR was rebased on another that is not yet merged; therefore, there is no need to review this one, as it will later be rebased on master.

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+ rebase-ff
@robodoo override=ci/security
Reason:
- `window.location`: redirection to a hard-coded, serverside-computed, URL (same as before).
Link to ci/security build (master): https://runbot.odoo.com/runbot/build/65153922

**Pull request closed without merging**